core/http/client/src/main/java/org/eclipse/rdf4j/http/client/SPARQLProtocolSession.java

```diff
@@ -685,15 +688,24 @@ protected void getTupleQueryResult(HttpUriRequest method, TupleQueryResultHandle
 						.orElseThrow(() -> new RepositoryException(
 								"Server responded with an unsupported file format: " + mimeType));
 
-				// Check if we can pass through to the output stream directly
+				// Check if we can pass through to the writer directly
 				if (handler instanceof TupleQueryResultWriter) {
 					TupleQueryResultWriter tqrWriter = (TupleQueryResultWriter) handler;
 					if (tqrWriter.getTupleQueryResultFormat().equals(format)) {
-						OutputStream out = tqrWriter.getOutputStream().orElse(null);
-						if (out != null) {
-							InputStream in = response.getEntity().getContent();
-							IOUtils.copy(in, out);
-							return;
+						if (tqrWriter instanceof CharSink) {
```

Comment:
I think the following block is quite long and a bit hard to read. Would it be possible to split it up with some private methods? Or replacing the `instanceof` conditionals with use of the Strategy/Visitor Pattern? I tend to prefer it to get better compile-time checks.

Reply:
Yeah good point, not sure about a visitor but I'll DRY it up a bit at least.

Reply:
Much better, even if I try to avoid `instanceof` - if possible.